text/0000-rustdoc-json.md

```diff
@@ -0,0 +1,1080 @@
+                      ]
+                    }
+                  },
+                  "param_names": null,
+                  "is_generic": false
+                }
+              },
+              "generic_params": [],
+              "modifier": "none"
+            }
+          }
+        ]
+      }
+    ]
+  ],
+  "output": {
+    "kind": "never"
+  }
+}
+```
+#### Generic Args
+```rust
+pub trait MyTrait<'a, T> {
+    type Item;
+    type Other;
+}
+
+pub fn generic_args<'a>(x: impl MyTrait<'a, i32, Item = u8, Other = f32>) {
+    unimplemented!()
+}
+```
+
+```json
+"decl": {
+  "inputs": [
+    [
+      "x",
+      {
+        "kind": "impl_trait",
+        "inner": [
+          {
+            "trait_bound": {
+              "trait": {
+                "kind": "resolved_path",
+                "inner": {
+                  "name": "MyTrait",
+                  "id": "0:11",
+                  "args": {
+                    "angle_bracketed": {
+                      "args": [
+                        {
+                          "lifetime": "'a"
+                        },
+                        {
+                          "type": {
+                            "kind": "primitive",
+                            "inner": "i32"
+                          }
+                        }
+                      ],
+                      "bindings": [
+                        {
+                          "name": "Item",
+                          "binding": {
+                            "equality": {
+                              "kind": "primitive",
+                              "inner": "u8"
+                            }
+                          }
+                        },
+                        {
+                          "name": "Other",
+                          "binding": {
+                            "equality": {
+                              "kind": "primitive",
+                              "inner": "f32"
+                            }
+                          }
+                        }
+                      ]
+                    }
+                  },
+                  "param_names": null,
+                  "is_generic": false
+                }
+              },
+              "generic_params": [],
+              "modifier": "none"
+            }
+          }
+        ]
+      }
+    ]
+  ],
+  "output": null
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- By supporting JSON output for `rustdoc`, we should consider how much it should mirror the
+  internal structures used in `rustdoc` and in the compiler. Depending on how much we want to
+  stabilize, we could accidentally stabilize the internal structures of `rustdoc`. We have tried
+  to avoid this by introducing a mirror of `rustdoc`'s AST types which exposes as few compiler
+  internals as possible by stringifying or not including certain fields.
+- Adding JSON output adds *another* thing that must be kept up to date with language changes,
```

Comment:
What would it take to generate rustdoc's current HTML from a JSON IR? It could be a way to ensure that changes are applied uniformly.

Reply:
If we wanted to use the JSON output as the "source of truth" and make the HTML renderer consume that, we'd just slap serde derives on the `clean` types so that we could round trip them rather than creating another set of types. The issue then is that you probably wouldn't want to make that JSON a public interface since it would have to expose some compiler internal types that the HTML renderer happens to use, and it would still be completely unstable.

I think it's worth keeping the JSON types as a separate representation because they have a different purpose than the `clean` types. Since they're supposed to be a relatively stable public interface, they don't need to be updated to add support every time an unstable feature is added to the compiler. It also allows us to drop some information for the sake of making the JSON easier to handle.